File: docs/validation/python/app/scenarios/s05_binary_roundtrip.py
import hashlib
import tempfile
from pathlib import Path
import numpy as np
from machine_config import MachineConfigReader, MachineConfigWriter

"""
S-05: Full binary round-trip: read, write, re-read, verify binary data preserved with correction hash
ID:          S-05
Title:       Copy config with binary data, verify correction hash unchanged
Category:    happy-path
Layer:       writer + reader (copy-hdf5 path)
Precondition: fixtures/reference_config.h5
Action:      Read with includeBinary=true → write to temp → read temp with
             includeBinary=true → compare SHA-256 of correction_data bytes
Expected:    SHA-256 of correction_data for train 0 identical before and after.
             SHA-256 of inverse_correction_data for train 0 identical before and after.
             file_size of scan_field_correction_file unchanged.
Rationale:   Silent precision loss in binary data is undetectable without hash comparison.
             This is the same check the cross_check Phase 4 performs — verify it
             also passes through the standalone app.
"""

def run(fixtures_dir: str, real_dir: str) -> tuple[bool, str]:
    path = Path(fixtures_dir) / "reference_config.h5"
    reader = MachineConfigReader(path)
    cfg = reader.parse()

    cd_before = reader.get_correction_data(0)
    icd_before = reader.get_inverse_correction_data(0)

    with tempfile.NamedTemporaryFile(suffix=".h5", delete=False) as f:
        tmp = f.name

    MachineConfigWriter(cfg).write(tmp)
    reader2 = MachineConfigReader(tmp)

    cd_after = reader2.get_correction_data(0)
    icd_after = reader2.get_inverse_correction_data(0)

    if not np.array_equal(cd_before, cd_after, equal_nan=True):
        h_before = hashlib.sha256(cd_before.tobytes()).hexdigest()[:16]
        h_after = hashlib.sha256(cd_after.tobytes()).hexdigest()[:16]
        return False, f"correction_data mismatch: {h_before}... → {h_after}..."
    if not np.array_equal(icd_before, icd_after, equal_nan=True):
        return False, "inverse_correction_data mismatch after roundtrip"

    cd_hash = hashlib.sha256(cd_before.tobytes()).hexdigest()
    return True, f"correction_data preserved: SHA-256={cd_hash}"

File: docs/validation/go/app/scenarios/s05_binary_roundtrip.go
package scenarios

// S-05: Full binary round-trip with correction hash verification
//
// ID:       S-05
// Action:   Read (with binary) -> write to temp -> read temp -> compare
//           SHA-256 of correction_data / inverse_correction_data.
// Rationale: silent precision loss in binary data is undetectable without a
//           hash comparison.

import (
	"fmt"
	"os"
	"path/filepath"

	mc "machine-config-go"
)

func RunS05BinaryRoundtrip(fixturesDir, _ string) (bool, string) {
	path := filepath.Join(fixturesDir, "reference_config.h5")
	reader := mc.NewReader(path)

	// ParseWithOptions(IncludeBinary: true), not Parse(): Write()ing a config
	// read via plain Parse() would omit the correction grid entirely, since
	// the scalars-only path never loads it.
	cfg, err := reader.ParseWithOptions(mc.ParseOptions{IncludeBinary: true})
	if err != nil {
		return false, fmt.Sprintf("ParseWithOptions failed: %v", err)
	}

	cdBefore, err := reader.GetCorrectionData(0)
	if err != nil {
		return false, fmt.Sprintf("GetCorrectionData failed: %v", err)
	}
	icdBefore, err := reader.GetInverseCorrectionData(0)
	if err != nil {
		return false, fmt.Sprintf("GetInverseCorrectionData failed: %v", err)
	}

	tmpPath, err := tempH5("mcl_go_s05")
	if err != nil {
		return false, fmt.Sprintf("tempfile: %v", err)
	}
	defer os.Remove(tmpPath)

	if err := mc.NewWriter().Write(cfg, tmpPath); err != nil {
		return false, fmt.Sprintf("write failed: %v", err)
	}

	reader2 := mc.NewReader(tmpPath)
	cdAfter, err := reader2.GetCorrectionData(0)
	if err != nil {
		return false, fmt.Sprintf("readback GetCorrectionData failed: %v", err)
	}
	icdAfter, err := reader2.GetInverseCorrectionData(0)
	if err != nil {
		return false, fmt.Sprintf("readback GetInverseCorrectionData failed: %v", err)
	}

	if !bitwiseEqual(cdBefore.Data, cdAfter.Data) {
		return false, fmt.Sprintf("correction_data mismatch: %s... -> %s...",
			sha256Hex(cdBefore.Data)[:16], sha256Hex(cdAfter.Data)[:16])
	}
	if !bitwiseEqual(icdBefore.Data, icdAfter.Data) {
		return false, "inverse_correction_data mismatch after roundtrip"
	}

	return true, fmt.Sprintf("correction_data preserved: SHA-256=%s", sha256Hex(cdBefore.Data))
}
